### DISCUSSION TITLE
Homogenization of linear elastic materials

### DISCUSSION THREAD

**Discussion by ghost** [kinnala/scikit-fem#1093] [Q&A]
Hi, sckit-fem team)
Сould you present a quick example considering homogenization of linear elastic materials, like the one shown
https://comet-fenics.readthedocs.io/en/latest/demo/periodic_homog_elas/periodic_homog_elas.html?

Thanks in advance for your reply

**Comment by ghost** [comment]
Hi, kinnala) I'm trying to solve a similar problem as indicated in the link above. My problem is based on MeshHex1 and I use ElementHex1 wrapped in ElementVector.
I need to add another strain tensor to the tensor T in the stress-strain function.
```
def linear_stress(Lambda=1., Mu=1.):

    def C(T):
        """Linear-elastic stress-strain relationship."""
        return 2. * Mu * T + Lambda * eye(trace(T), T.shape[0])

    return C
```
For instance, another tensor looks like this:
```math
 \begin{pmatrix}
  1 & 0 & 0 \\
  0 & 0 & 0 \\
  0 & 0 & 0
 \end{pmatrix}

```
In this case, will the stress-strain function look like below or not?
```
def linear_stress(Lambda=1., Mu=1.):

    def C(T):
        T[0, 0] += 1.
        return 2. * Mu * T + Lambda * eye(trace(T), T.shape[0])

    return C
```

**Reply by kinnala** [reply]
Yes.

Sorry for not answering the original question. I believe we have all the required functionality, but it is somewhat new topic for me so I would need to read it carefully in order to come up with an example.

**Reply by ghost** [reply]
Hi, everyone) Here's a little question. What does the strain-stress relationship look like in the scikit-fem?

**Reply by kinnala** [reply]
The library does not enforce any specific stress-strain relationship, the only one we have implemented as an example for others to use is here: https://github.com/kinnala/scikit-fem/blob/02a80c3f99af6410b353af0842ccfcf80c4f0cc2/skfem/models/elasticity.py#L39

**Reply by ghost** [reply]
Hi, everyone) I don't want to set up a separate topic, so I'll leave a couple of sentences about Mesh.with_subdomains.

The current signature is:
```python
def with_subdomains(self, subdomains: Dict[str, Callable[[ndarray], ndarray]]):
"""Return a copy of the mesh with named subdomains.

        Parameters
        ----------
        boundaries
            A dictionary of lambda functions with the names of the subdomains
            as keys.  The midpoint of the element should return ``True`` for
            the corresponding lambda function if the element belongs to the
            subdomain.

        """
        return replace(
            self,
            _subdomains={
                **({} if self._subdomains is None else self._subdomains),
                **{name: (self.elements_satisfying(test)
                          if callable(test) else test)
                   for name, test in subdomains.items()},
            },
        )
```
1.  In the docstring, 'boundaries' should be replaced with 'subdomains'.
2. This function is similar to with_boundaries, so:
2.1 Perhaps the following signature should be defined:
```python
def with_subdomains(self, subdomains: Dict[str, Union[Callable[[ndarray], ndarray],  ndarray]]):
```
2.2 Make changes to the Parameters description:
"
Parameters
        ----------
        subdomains
            A dictionary of index arrays or lambda functions with the names of the subdomains
            as keys.
"

**Comment by ghost** [comment]
Hi, kinnala) I don't want to set up a separate topic again, so I'll leave a couple suggestions here.
I didn't find any example related to nonlinear time-dependent ODE in the current list (maybe I just wasn't looking carefully).
But you have such an example, available at the link
https://scicomp.stackexchange.com/questions/35940/1d-fem-for-nonlinear-diffusion-coefficient
Could you please revise it to the current version of skfem and post it here. I think this would be useful for everyone using your application.
